engines/scumm/he/basketball/ai.cpp

```diff
@@ -0,0 +1,247 @@
+
+	return 1;
+}
+
+int Basketball::numOpponentsInCone(int team, float widthDistanceRatio, const U32FltVector2D &end, const U32FltVector2D &focus) {
+	// The end out parameter is the end point of the center line of the cone,
+	// and widthDistanceRatio defines the width of the cone = width at dist 1
+	int count = 0;
+
+	Line2D line(focus, end);
+
+	Common::Array<CCollisionPlayer> *playerList = (team == TEAM_HOME) ? &_vm->_basketball->_court->_homePlayerList : &_vm->_basketball->_court->_awayPlayerList;
+
+	for (Common::Array<CCollisionPlayer>::iterator currentPlayer = playerList->begin(); currentPlayer != playerList->end(); ++currentPlayer) {
+		U32FltVector2D point = line.projectPoint(currentPlayer->center);
+
+		if (focus.distance2(currentPlayer->center) < (point.distance2(focus) * widthDistanceRatio * widthDistanceRatio) && line.inBetween(point, focus, end)) {
+			count++;
+		}
+	}
+
+	return count;
+}
+
+static bool isPointInBounds(const U32FltVector2D &point) {
+	return ((point.x > 0) &&
+			(point.x < MAX_WORLD_X) &&
+			(point.y > 0) &&
+			(point.y < MAX_WORLD_Y));
+}
+
+float Basketball::congestion(U32FltVector2D pos, bool ignore, int whichPlayer) {
+	float congestion = 0.0F;
+
+	for (int team = TEAM_HOME; team <= TEAM_AWAY; ++team) {
+		Common::Array<CCollisionPlayer> *playerList = (team == TEAM_HOME) ? &_court->_homePlayerList : &_court->_awayPlayerList;
+
+		for (Common::Array<CCollisionPlayer>::iterator currentPlayer = playerList->begin(); currentPlayer != playerList->end(); ++currentPlayer) {
+			if (currentPlayer->_playerIsInGame && (ignore && !(currentPlayer->_objectID == whichPlayer))) {
+				float distance2 = pos.distance2(currentPlayer->center);
+
+				if (distance2 == 0.0F)
+					return 3.402823466e+38F;
+
+				congestion += (1 / distance2);
+			}
+		}
+	}
+
+	return congestion;
+}
+
+int LogicHEBasketball::u32_userGetOpenSpot(int whichPlayer, U32FltVector2D upperLeft, U32FltVector2D lowerRight, U32FltVector2D passer, bool attract, U32FltVector2D attractPoint) {
+	int xGranularity = 5;
+	int yGranularity = 5;
+
+	int rectWidth = fabs(upperLeft.x - lowerRight.x);
+	int rectHeight = fabs(upperLeft.y - lowerRight.y);
+
+	float xMesh = rectWidth / (xGranularity + 1);
+	float yMesh = rectHeight / (yGranularity + 1);
+
+	float startX = upperLeft.x + xMesh / 2;
+	float startY = upperLeft.y + yMesh / 2;
+
+	float x = startX, y = startY;
+
+	float bestCongestion = 3.402823466e+38F;
+
+	U32FltVector2D bestPoint, point;
+
+	float tmp;
+
+	for (int ii = 0; ii < xGranularity; ii++) {
+		for (int jj = 0; jj < yGranularity; jj++) {
+			tmp = _vm->_basketball->congestion(point = U32FltVector2D(x, y), true, whichPlayer);
+
+			if (attract) {
+				tmp -= (4 * (1 / point.distance2(attractPoint)));
+			}
+
+			if (tmp < bestCongestion &&
+				isPointInBounds(point)) {
+				bestCongestion = tmp;
+				bestPoint = point;
+			}
+
+			y += yMesh;
+		}
+
+		x += xMesh;
+		y = startY;
+	}
+
+	if (bestCongestion == 3.402823466e+38F) {
```

Comment:
Perhaps this magic value should be added to a define?

Reply:
This actually was:
`#define FLT_MAX          3.402823466e+38F        // max value`
do we have something already which won't fail on some platform while compiling correctly on others?

Reply:
I don't see how changing the value to be a define will have an impact on the platform it's used. Do you have a specific use case in mind?

Reply:
The FLT_MAX define is a standard define like INT_MAX, etc. The thing is not all platform have it and I keep forgetting what we should be using in ScummVM instead. Paging @sev- 😛

Reply:
We define FLT_MAX if it's missing in math/utils.h:
https://github.com/scummvm/scummvm/blob/721dc4fbc6334ef7a153f2d0ae371fac260f7154/math/utils.h#L32

So, it should be OK to use

Reply:
@AndywinXp note that you need to include `math/utils.h` to use FLT_MAX

Reply:
Thank you, I missed that. Fixed!